clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -460,6 +401,79 @@ llvm::Type *CodeGenTypes::ConvertFunctionTypeInternal(QualType QFT) {
+  // Once we update to the newest version of the spec - this should be updated.
+  assert((TemplateArgs.size() == 5 || TemplateArgs.size() == 6) &&
+         "Wrong JointMatrixINTEL template parameters number");
+  // This is required to represent optional Optional
+  // 'Component Type Interpretation' parameter
+  using ParamsType =
+      typename std::conditional<NeedTypeInterpret, SmallVector<unsigned, 6>,
+                                SmallVector<unsigned, 5>>::type;
+  ParamsType Params;
+  if constexpr (NeedTypeInterpret)
+    Params = { 0, 0, 0, 0, 0, Val};
+  else
+    Params = { 0, 0, 0, 0, 0};
+  for (size_t I = 1; I != TemplateArgs.size(); ++I) {
+    assert(TemplateArgs[I].getKind() ==
+           TemplateArgument::Integral &&
+           "Wrong JointMatrixINTEL template parameter");
+    Params[I - 1] = TemplateArgs[I].getAsIntegral().getExtValue();
+  }
+  return llvm::TargetExtType::get(
+      CompTy->getContext(), "spirv.JointMatrixINTEL", {CompTy}, Params);
+}
+
+/// ConvertSYCLJointMatrixINTELType - Convert SYCL joint_matrix type
+/// which is represented as a pointer to a structure to LLVM extension type
+/// with the parameters that follow SPIR-V JointMatrixINTEL type.
+/// The expected representation is:
+/// target("spirv.JointMatrixINTEL", %element_type, %rows%, %cols%, %scope%,
+///        %use%, (optional) %element_type_interpretation%)
+llvm::Type *CodeGenTypes::ConvertSYCLJointMatrixINTELType(RecordDecl *RD) {
+  auto *TemplateDecl = cast<ClassTemplateSpecializationDecl>(RD);
+  ArrayRef<TemplateArgument> TemplateArgs =
+      TemplateDecl->getTemplateArgs().asArray();
+  assert(TemplateArgs[0].getKind() == TemplateArgument::Type &&
+        "1st JointMatrixINTEL template parameter must be type");
+  llvm::Type *CompTy = ConvertType(TemplateArgs[0].getAsType());
+
+  // Per JointMatrixINTEL spec the type can have an Optional
+  // 'Component Type Interpretation' parameter. We should emit it in case
+  // if on SYCL level joint matrix accepts 'bfloat16' or 'tf32' objects as
+  // matrix's components. Yet bfloat16 should be represented as 'int16' and
+  // 'tf32' as 'float' types.
+  if (CompTy->isStructTy()) {
+    StringRef LlvmTyName = CompTy->getStructName();
+    // Emit half/int16/float for sycl[::*]::{half,bfloat16,tf32}
+    if (LlvmTyName.startswith("class.sycl::") ||
+        LlvmTyName.startswith("class.__sycl_internal::"))
+      LlvmTyName = LlvmTyName.rsplit("::").second;
+    if (LlvmTyName == "half") {
+      CompTy = llvm::Type::getHalfTy(getLLVMContext());
+      return getJointMatrixINTELExtType(CompTy, TemplateArgs);
+    } else if (LlvmTyName == "tf32") {
+      CompTy = llvm::Type::getHalfTy(getLLVMContext());
+      // 'tf32' interpretation is mapped to '0'
+      return getJointMatrixINTELExtType<true>(CompTy, TemplateArgs, 0);
+    } else if (LlvmTyName == "bfloat16") {
+      CompTy = llvm::Type::getInt16Ty(getLLVMContext());
```

Comment:
LLVM has a `bfloat` type, so surely it would make more sense to use that instead of `i16`? Or is there something about the JointMatrixINTEL spec that I'm missing?

Reply:
The thing is that in SPIR-V we don't have bfloat type (yet?). Instead we have conversion instructions and using `short` value as bfloat16 storage. Same for SYCL - here `bfloat16` class is a wrapper around int16 `storage`. So for the consistency `int16` is used here as well, though there is no other reason not to reuse LLVM's `bf16` (just because we still will replace it with `int16` in SPIR-V. So I lean towards appying this suggestion.

Reply:
Actually, lets keep i16 to be aligned with SPIR-V spec. Once we have proper SPIR-V bf16 type we can start generating in by DPC++ compiler either here or elsewhere.